codec/src/codec.rs

```diff
@@ -1,63 +1,156 @@
-//! Core codec traits and implementations
+//! Core traits for encoding and decoding.
 
 use crate::error::Error;
 use bytes::{Buf, BufMut, BytesMut};
+use std::ops::RangeBounds;
 
-/// Trait for types that can be encoded to and decoded from bytes
-pub trait Codec: Sized {
-    /// Encodes this value to a writer.
+/// Marker trait for types that can be used as configuration during decoding.
+///
+/// Configuration is primarily used with the [`Read`] trait to pass parameters (like size limits)
+/// needed to safely decode untrusted data. Types implementing `Config` must also be
+/// `Clone + Send + 'static`.
+///
+/// Use the unit type `()` if no configuration is required for a specific [`Read`] implementation.
+pub trait Config: Clone + Send + 'static {}
+
+// Automatically implement `Config` for matching types.
+impl<T: Clone + Send + 'static> Config for T {}
+
+/// A marker trait for a [`Config`] type that is also a [`RangeBounds<usize>`].
+///
+/// This is often used to configure length limits for variable-length collections like `Vec<T>` or
+/// `Bytes`.
+pub trait RangeConfig: Config + RangeBounds<usize> {}
+
+// Automatically implement `RangeConfig` for matching types.
+impl<T: Config + RangeBounds<usize>> RangeConfig for T {}
+
+/// Trait for types with a known, fixed encoded size.
+///
+/// Implementing this trait signifies that the encoded representation of this type *always* has the
+/// same byte length, regardless of the specific value.
+///
+/// This automatically provides an implementation of [`EncodeSize`].
+pub trait FixedSize {
+    /// The size of the encoded value (in bytes).
+    const SIZE: usize;
+}
+
+/// Trait for types that can provide their encoded size in bytes.
+///
+/// This must be implemented by all encodable types. For types implementing [`FixedSize`], this
+/// trait is implemented automatically. For variable-size types, this requires calculating the size
+/// based on the value.
+pub trait EncodeSize {
```

Comment:
nit: I think the `encode_` aspect is implied...wondering if we just call it `fn size()`?

Reply:
I agree that when looking at this code, the `encode_` aspect is obvious and redundant. However, remember that these types will be used externally. In this case, it will be much clearer what `encode_size()` means and prevent bugs. For example, consider a collection, say a map that is encodable. We don't want someone to accidentally call `size()` when they mean to call `len()`. By calling it `encode_size()` we remove this issue and make it more clear where the function is coming from